### ISSUE TITLE
Wix Fragments  support

### ISSUE THREAD

**Issue by jvmlet** [i-net-software/SetupBuilder#76]
It would be great to support  WIX external  **Fragment** documents  in  msi dsl to be used in custom template.

**Comment by Horcrux7** [comment]
What do you want to do? How should look the the Gradle API for it? You can use all WIX features if you use a template with your additional features.

The SetupBuilder plugin has a cross platform target.  I can not see which benefit such low level feature has for the other platforms.

**Comment by jvmlet** [comment]
I'm trying to have external WIX document with **Fragment** node that defines Dialogs.
The gradle dsl will look something like this  :

````groovy
msi {
 externals= [fragment1.wxs,fragment2.wxs]
}
````
The task will compile the externals with and run the main as described [here ](https://www.firegiant.com/wix/tutorial/user-interface/new-link-in-the-chain/):
````
candle.exe SampleWixUIAddDlg.wxs UserRegistrationDlg.wxs
light.exe -ext WixUIExtension -out SampleWixUIAddDlg.msi SampleWixUIAddDlg.wixobj UserRegistrationDlg.wixobj
````

**Comment by Horcrux7** [comment]
```
msi {
 externals= [fragment1.wxs,fragment2.wxs]
}
```
I see not what this code can do what you can't already do with a custom template. You can generate your template in a depending task before.

**Comment by jvmlet** [comment]
What do you mean by generate the template ? My dialogs  are defined in external document and WIX doesn't allow to have both **Frament** and **Product** to be defined in the same document.
Can you please give a pseudo-dsl-code of what you mean ?

**Comment by Horcrux7** [comment]
> WIX doesn't allow to have both Frament and Product

Then I does not know how SetupBuilder should implement this. The `Product` Element is a core element in SetupBuilder.

> Can you please give a pseudo-dsl-code of what you mean ?

This is also my question.

The current wxs template can look like:
```
<?xml version="1.0" encoding="UTF-8"?>
<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi">
    <Product>
     ...
    </Product>
</Wix>
```
Then you have a file with a fragment:
```
<Fragment>
  ...
</Fragment>
```
This both should be merged which result:
```
<?xml version="1.0" encoding="UTF-8"?>
<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi">
    <Product>
     ...
    </Product>
    <Fragment>
        ...
    </Fragment>
</Wix>
```
Such a merge can you realize with the template.wxs. If you need any other then you must described what you want. How should look the resulting wsx file. A PR can be a solution to do this. We does not know all the features of WIX. Its sounds like that you understand many more from WIX as we. Please describe completely what you want that a Java developer can understand it. You can attached sample files if this help.

**Comment by jvmlet** [comment]
@Horcrux7 ,
As I already mentioned, wix xsd schema doesn't allow both **Product** and **Fragment**  elements to be used in the same document,

 [This guide](https://www.firegiant.com/wix/tutorial/user-interface/new-link-in-the-chain/) describes how the external  fragment is used.
Given the below gradle dsl 

````
msi {
 externals= [fragment1.wxs,fragment2.wxs]
}
````
you will run below WIX commands from gradle task:

````
candle.exe template.wxs fragment1.wxs fragmen2.wxs
light.exe -ext WixUIExtension -out project-name.msi  template.wixobj   fragment1.wixobj  fragment2.wixobj
````

Hope it's  more clear now

**Comment by Horcrux7** [comment]
The link does not described what to do with with all the other data. It described only fragments. This does not help to understand it. Also if you post it a third time.

Please make a PR. Then we can look into. Currently we does not understand how this should be work with the existing code base.

**Comment by jvmlet** [comment]
@Horcrux7 , would you please  review the commit.

**Comment by Horcrux7** [comment]
Merged! It will be nice if you can add API doc to the new methods in the MSI class.

**Comment by jvmlet** [comment]
Added in PR  #84

**Comment by Horcrux7** [comment]
Can we close this issue now?

**Comment by jvmlet** [comment]
Sure, would you please release the SetupBulder ?

**Comment by Horcrux7** [comment]
We wait on a fix for #82 before we want release.
